src/Microsoft.DotNet.Build.Tasks.Feed/BlobFeedAction.cs

```diff
@@ -277,12 +329,80 @@ private AzureFileSystem GetAzureFileSystem()
             return fileSystem;
         }
 
-        private async Task<bool> PushAsync(IEnumerable<string> items, bool allowOverwrite)
+        private async Task<bool> PushAsync(
+            IEnumerable<string> items,
+            bool allowOverwrite,
+            bool passIfExistingItemIdentical)
         {
             LocalSettings settings = GetSettings();
             AzureFileSystem fileSystem = GetAzureFileSystem();
-            bool result = await PushCommand.RunAsync(settings, fileSystem, items.ToList(), allowOverwrite, skipExisting: false, log: new SleetLogger(Log));
-            return result;
+            SleetLogger log = new SleetLogger(Log);
+
+            if (!allowOverwrite && passIfExistingItemIdentical)
+            {
+                var context = new SleetContext
+                {
+                    LocalSettings = settings,
+                    Log = log,
+                    Source = fileSystem,
+                    Token = CancellationToken
+                };
+                context.SourceSettings = await FeedSettingsUtility.GetSettingsOrDefault(
+                    context.Source,
+                    context.Log,
+                    context.Token);
+
+                var flatContainer = new FlatContainer(context);
+
+                var packageIndex = new PackageIndex(context);
+
+                var packagesToPush = new List<string>();
+
+                // Check packages sequentially: Task.WhenAll caused IO exceptions in Sleet.
+                foreach (var item in items)
+                {
+                    bool? identical = await IsPackageIdenticalOnFeedAsync(
+                        item,
+                        packageIndex,
+                        context.Source,
+                        flatContainer,
+                        log);
+
+                    if (identical == true)
+                    {
+                        Log.LogMessage(
+                            MessageImportance.Normal,
+                            "Package exists on the feed, and is verified to be identical. " +
+                            $"Skipping upload: '{item}'");
+                    }
+                    else if (identical == false)
+                    {
+                        Log.LogError(
+                            "Package exists on the feed, but contents are different. " +
+                            $"Upload failed: '{item}'");
+                    }
+                    else
+                    {
+                        packagesToPush.Add(item);
+                    }
+                }
+
+                items = packagesToPush;
+
+                if (!items.Any())
+                {
+                    Log.LogMessage("After skipping idempotent uploads, no items need pushing.");
+                    return true;
+                }
+            }
+
+            return await PushCommand.RunAsync(
+                settings,
+                fileSystem,
+                items.ToList(),
```

Comment:
I would return packagesToPublish here instead of items to avoid the ToList() call. In that case you don't need line 390 and need to update 392

Reply:
Agreed, didn't notice this ToList. Fixing.